util/no_os_lf256fifo.c

```diff
@@ -0,0 +1,64 @@
+#include <errno.h>
+#include "no_os_lf256fifo.h"
+
+struct lf256fifo {
+    char * data; // pointer to memory area where the buffer will be stored
+	uint8_t ffilled; // the index where the data starts (\note Read it as 'first-filled-slot')
+	uint8_t fempty; // the index where empty/non-used area starts (\note Read it as 'first-empty-slot')
+};
+
+int lf256fifo_init(struct lf256fifo **fifo)
+{
+	if (fifo == NULL)
+		return -EINVAL;
+
+	struct lf256fifo *b = calloc(1, sizeof(struct lf256fifo));
+	if (b == NULL)
+		return -ENOMEM;
+
+	b->data = calloc(1, 256);
+	if (b->data == NULL)
+		return -ENOMEM;
+
+	*fifo = b;
+
+	return 0;
+}
+
+bool lf256fifo_is_full(struct lf256fifo *fifo)
+{
+	return (fifo->fempty + 1) == fifo->ffilled; // intended overflow at 256 (data size is 256)
+}
+
+bool lf256fifo_is_empty(struct lf256fifo *fifo)
+{ 
+	return fifo->fempty == fifo->ffilled;
+}
+
+int lf256fifo_read(struct lf256fifo * fifo, char *c)
+{
+	if (lf256fifo_is_empty(fifo))
+		return -1; // buffer empty
+
+	*c = fifo->data[fifo->ffilled];
+	fifo->ffilled++; // intended overflow at 256 (data size is 256)
```

Comment:
This is actually a fifo implemented as a circular buffer on a fixed 256 char memory area.
ffilled and fempty are intentionately uint8_t so that when we do ++ on them and they overflow, they point to the beginning of the memory area. This is all intentional and proven to work in tests.

Reply:
regarding circular_buffer, I had to make this specific fifo because of concurrency issues with circular_buffer when using it from normal user context + interrupt context.